### PR TITLE
ADC configuration for phototransistors 

### PR DESCRIPTION
**Pull request by cua-cua**
-ADC1
-4 injected channels (one by phototransistor sensor)
-Scan mode triggered by sw
Issue #55

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -288,31 +291,67 @@ static uint32_t read_encoder_right(void)
...
+}
+
+
+/**
  * @brief Initial setup and infinite wait.
  */
 int main(void)
 {
 	int j = 0;
+	uint16_t adc_read = 0;
```
Comment (Peque) [inline src/main.c]: Maybe this variable declaration could be removed?

File: src/main.c [hunk 2]
```diff
@@ -288,31 +291,67 @@ static uint32_t read_encoder_right(void)
...
 
 /**
+ * @brief Setup for adc: Four injected channels on ADC1.
+ */
+static void setup_adc(void)
+{
+	int i;
+	uint8_t channel_array[4];
+
+	/* Make sure the ADC doesn't run during config. */
```
Comment (Peque) [inline src/main.c]: I would rather include these comments in the docstring.

File: src/main.c [hunk 3]
```diff
@@ -288,31 +291,67 @@ static uint32_t read_encoder_right(void)
 
 
 /**
+ * @brief Setup for adc: Four injected channels on ADC1.
```
Comment (Peque) [inline src/main.c]: Maybe add a reference/note in the docstring referring to the manual.

File: src/main.c [hunk 4]
```diff
@@ -40,6 +40,9 @@ static void setup_clock(void)
 
 	/* Alternate functions */
 	rcc_periph_clock_enable(RCC_AFIO);
+
+	/*Adc*/
```
Comment (Peque) [inline src/main.c]: ADC in caps?

File: src/main.c [hunk 5]
```diff
@@ -288,31 +291,67 @@ static uint32_t read_encoder_right(void)
...
+{
+	int i;
+	uint8_t channel_array[4];
+
+	/* Make sure the ADC doesn't run during config. */
+	adc_power_off(ADC1);
+
+	adc_enable_scan_mode(ADC1);
+	adc_set_single_conversion_mode(ADC1);
+	/* We want to start the injected conversion in SW*/
```
Comment (Peque) [inline src/main.c]: I would rather include these comments in the docstring.

File: src/main.c [hunk 6]
```diff
@@ -288,31 +291,67 @@ static uint32_t read_encoder_right(void)
...
+	 */
+	channel_array[0] = 4;
+	channel_array[1] = 5;
+	channel_array[2] = 6;
+	channel_array[3] = 7;
+	adc_set_injected_sequence(ADC1, 4, channel_array);
+	adc_power_on(ADC1);
+
+	/* Wait for ADC starting up. */
+	for (i = 0; i < 800000; i++)    /* Wait a bit. */
```
Comment (Peque) [inline src/main.c]: Maybe we can remove the "wait a bit" comment (repeated).

File: src/main.c [hunk 7]
```diff
@@ -288,31 +291,67 @@ static uint32_t read_encoder_right(void)
...
+		__asm__("nop");
```
Comment (Peque) [inline src/main.c]: Isn't there a function that waits for the ADC to be started? (in libopencm3, I mean)

File: src/main.c [hunk 8]
```diff
@@ -288,31 +291,67 @@ static uint32_t read_encoder_right(void)
 
 
 /**
+ * @brief Setup for adc: Four injected channels on ADC1.
+ */
+static void setup_adc(void)
+{
+	int i;
+	uint8_t channel_array[4];
```
Comment (Peque) [inline src/main.c]: I would rather put here the initialization of the channel array. In example:

    uint8_t channel_sequence[] = [ADC_CHANNELX, ADC_CHANNELY, ... ADC_CHANNELZ];

Pseudocode... :joy:

File: src/main.c (hunk 6 above)
Comment (Peque) [reply src/main.c]: Maybe we could say "wait about 3 microseconds". We could add a note in the docstring pointing to the `adc_power_on` documentation (where they suggest to use this wait).

By the way, how much are we waiting with that 800000?

Comment (cua-cua) [reply src/main.c]: I took that value from libopencm3 official examples for ADC (stm32/f1/lisa-m-2/adc_injec etc).
I guess that, if the clock is at 72 MHz, the nop takes 1 cycle, and it is called 800000. 
(800000/(72*10^6) = 0.01 s)

File: src/main.c (hunk 5 above)
Comment (cua-cua) [reply src/main.c]: Comments, comments dubidu, everything on docstring now :)

File: src/main.c (hunk 4 above)
Comment (cua-cua) [reply src/main.c]: ok

File: src/main.c (hunk 8 above)
Comment (cua-cua) [reply src/main.c]: parenthesis is the key

File: src/main.c (hunk 7 above)
Comment (cua-cua) [reply src/main.c]: I didn't find any :(

File: src/main.c (hunk 1 above)
Comment (cua-cua) [reply src/main.c]: yes! opened a new issue (#57)  to include static analysis on travis

File: src/main.c (hunk 6 above)
Comment (cua-cua) [reply src/main.c]: Comments updated ;)

**Comment by cua-cua** [comment]
Travis proposes to use ARRAY_SIZE(channel_sequence) from Linux Kernel. In my opinion, this should be disabled.

File: src/main.c [hunk 9]
```diff
@@ -40,6 +40,9 @@ static void setup_clock(void)
 
 	/* Alternate functions */
 	rcc_periph_clock_enable(RCC_AFIO);
+
+	/*ADC*/
```
Comment (Peque) [inline src/main.c]: Spaces around `/* ADC */`? :innocent:

File: src/main.c [hunk 10]
```diff
@@ -288,6 +291,46 @@ static uint32_t read_encoder_right(void)
 
 
 /**
+ * @brief Setup for adc: Four injected channels on scan mode for ADC1.
```
Comment (Peque) [inline src/main.c]: Just for consistency with other functions, could you add a blank line after the "brief" section? :blush:

Comment (Peque) [reply src/main.c]: Also, `ADC` should maybe go in caps. :stuck_out_tongue_winking_eye:

File: src/main.c [hunk 11]
```diff
@@ -288,6 +291,46 @@ static uint32_t read_encoder_right(void)
...
+ * - Initialize channel_sequence structure to map physical channels vs
+ * sw injected channels.
+ * - Power off the ADC to be sure that doesn't run during configuration.
+ * - Enable scan mode with single conversion mode triggered by software.
+ * - Configure the alignment (right) and the sample time (28.5 cycles of ADC
+ * clock)
+ * - Set injected sequence with channel_sequence structure
+ * - Power on the ADC and wait for ADC starting up (about 3 ms, See adc_power_on
+ * function from libopencm3)
+ * - Calibrate the ADC
```
Comment (Peque) [inline src/main.c]: Not very important, but to be consistent with docstring styles:

- Maybe we could avoid contractions? (i.e.: "doesn't", "vs", "sw"...)
- Maybe try to always end the list items with a dot?
- Maybe add 2 spaces when continuing a list item in a second line? (i.e.: like in `setup_pwm` docstring)

Do not know if there is a script to check that as well to put it in Travis... :joy:

File: src/main.c [hunk 12]
```diff
@@ -288,6 +291,46 @@ static uint32_t read_encoder_right(void)
...
 /**
+ * @brief Setup for adc: Four injected channels on scan mode for ADC1.
+ * - Initialize channel_sequence structure to map physical channels vs
+ * sw injected channels.
+ * - Power off the ADC to be sure that doesn't run during configuration.
+ * - Enable scan mode with single conversion mode triggered by software.
+ * - Configure the alignment (right) and the sample time (28.5 cycles of ADC
+ * clock)
+ * - Set injected sequence with channel_sequence structure
+ * - Power on the ADC and wait for ADC starting up (about 3 ms, See adc_power_on
```
Comment (Peque) [inline src/main.c]: Maybe this "See ..." could be removed from here and added at the end in a "@see ..."?

File: src/main.c [hunk 13]
```diff
@@ -288,6 +291,46 @@ static uint32_t read_encoder_right(void)
...
+ *
+ * @see Reference manual (RM0008) "Analog-to-digital converter" and in
+ * particular "Scan mode" section.
+ */
+static void setup_adc(void)
+{
+	int i;
+
+	uint8_t channel_sequence[4] = {ADC_CHANNEL4, ADC_CHANNEL5, ADC_CHANNEL6,
+	ADC_CHANNEL7};
```
Comment (Peque) [inline src/main.c]: What if we used something like...?

```c
uint8_t sequence_size = 4
uint8_t channel_sequence[sequence_size] = {...};

/* [...] */
adc_set_injected_sequence(ADC1, sequence_size, channel_sequence);
```

It is true that this way you need to change 2 lines if you want to change the sequence size, but at least they are together and, most important, the compiler will complain if you do it wrong. On the bright side, we do not need to use the `sizeof` later. :thinking:

File: src/main.c [hunk 14]
```diff
@@ -288,6 +291,46 @@ static uint32_t read_encoder_right(void)
...
+	adc_set_single_conversion_mode(ADC1);
+	adc_enable_external_trigger_injected(ADC1, ADC_CR2_JEXTSEL_JSWSTART);
+	adc_set_right_aligned(ADC1);
+	adc_set_sample_time_on_all_channels(ADC1, ADC_SMPR_SMP_28DOT5CYC);
+	adc_set_injected_sequence(ADC1,
+		sizeof(channel_sequence) / sizeof(channel_sequence[0]),
+		channel_sequence);
+	adc_power_on(ADC1);
+	for (i = 0; i < 800000; i++)
+		__asm__("nop");
```
Comment (Peque) [inline src/main.c]: Not very important now, but if in the future we do not know what to do...

https://github.com/Theseus/theseus/issues/59

:joy:

File: src/main.c (hunk 10 above)
Comment (cua-cua) [reply src/main.c]: ok

File: src/main.c (hunk 9 above)
Comment (cua-cua) [reply src/main.c]: Done

File: src/main.c (hunk 11 above)
Comment (cua-cua) [reply src/main.c]: Yes, I will try to include it on my editor

File: src/main.c (hunk 12 above)
Comment (cua-cua) [reply src/main.c]: See deleted

File: src/main.c (hunk 13 above)
Comment (cua-cua) [reply src/main.c]: Ok, good solution.

File: src/main.c (hunk 14 above)
Comment (cua-cua) [reply src/main.c]: good

**Comment by cua-cua** [comment]
Included gpio initialization for ADC pins, that I forgot last time. It was working anyway, I suppose that this configuration is the one by default.

**Comment by Peque** [comment]
Awesome! :+1: :blush:

Would you mind rewording the commit message?

Also, maybe we could agree on some general indentation rules for multi-line instructions. In this pull request we can find:

```c
/* Start after `(`, then break line and add 1 tab */
function_call(param0,
        param1, param2, param3);

/* Start after `{` then break line and do not add tab */
type variable[x] = {value0,
value1, value2};
```

What I have been using before in the code is:

```c
/* Start after `(`, then align following lines */
function_call(param0,
              param1, param2, param3);
```

We could also use (if, for example, first parameter does not even fit in the line with the function call):

```c
/* Start next line with one tab indentation */
really_long_function_call(
        really_long_parameter_1,
        really_long_parameter_2);
```

I do not think we need strict rules for that, but at least agree in some preferred way to try to keep the same style where possible. :wink:

**Comment by cua-cua** [comment]
Commit edited and clang-format applied ^^

**Pull request merged**